chia/cmds/cmd_classes.py

```diff
@@ -0,0 +1,322 @@
+        else:
+            return any([subclass.needs_context() for subclass in self.my_subclasses.values()])
+
+    def get_all_option_decorators(self) -> List[Callable[[SyncCmd], SyncCmd]]:
+        all_option_decorators: List[Callable[[SyncCmd], SyncCmd]] = self.my_option_decorators
+        for subclass in self.my_subclasses.values():
+            all_option_decorators.extend(subclass.get_all_option_decorators())
+        return all_option_decorators
+
+    def initialize_instance(self, **kwargs: Any) -> ChiaCommand:
+        kwargs_to_pass: Dict[str, Any] = {}
+        for kwarg_name in self.my_kwarg_names:
+            kwargs_to_pass[kwarg_name] = kwargs[kwarg_name]
+
+        for subclass_arg_name, subclass in self.my_subclasses.items():
+            kwargs_to_pass[subclass_arg_name] = subclass.initialize_instance(**kwargs)
+
+        return self.my_dataclass(**kwargs_to_pass)
+
+    def apply_decorators(self, cmd: SyncCmd) -> SyncCmd:
+        cmd_to_return = cmd
+        if self.needs_context():
+
+            def strip_click_context(func: SyncCmd) -> SyncCmd:
+                def _inner(ctx: click.Context, **kwargs: Any) -> None:
+                    context: Dict[str, Any] = ctx.obj if ctx.obj is not None else {}
+                    func(context=context, **kwargs)
+
+                return _inner
+
+            cmd_to_return = click.pass_context(strip_click_context(cmd_to_return))
+
+        for decorator in self.get_all_option_decorators():
+            cmd_to_return = decorator(cmd_to_return)
+
+        return cmd_to_return
+
+
+def _generate_command_parser(cls: Type[ChiaCommand]) -> _CommandParsingStage:
+    option_decorators: List[Callable[[SyncCmd], SyncCmd]] = []
+    kwarg_names: List[str] = []
+    subclasses: Dict[str, _CommandParsingStage] = {}
+    needs_context: bool = False
+
+    hints = get_type_hints(cls)
+    _fields = fields(cls)  # type: ignore[arg-type]
+
+    for _field in _fields:
+        field_name = _field.name
+        if (
+            isinstance(hints[field_name], type)
+            and hasattr(hints[field_name], "__command_helper__")
+            and hints[field_name].__command_helper__
+        ):
+            subclasses[field_name] = _generate_command_parser(hints[field_name])
+        else:
+            if field_name == "context":
+                if hints[field_name] != Context:
+                    raise ValueError("only Context can be the hint for variables named 'context'")
+                else:
+                    needs_context = True
+                    kwarg_names.append(field_name)
+                    continue
+
+            if "option_args" not in _field.metadata:
+                continue
+            option_args = _field.metadata["option_args"]
+
+            if "type" not in option_args:
+                origin = get_origin(hints[field_name])
+                if origin == collections.abc.Sequence:
+                    if "multiple" not in option_args or not option_args["multiple"]:
```

Comment:
Maybe preset the defaults somewhere like `option_args.setdefault("multiple", False)`.  Or something like `option_args = {"this": this_default, etc...}; option_args.update(provided_values)`.

Maybe the defaults should be pulled out of the click signatures?  Though sure, I can imagine they aren't always effectively available in the signature and that could turn into a ginormous rabbit hole.

Reply:
I did this for "multiple".  You could make an argument for "required" too but I think I wanted to force people to put `required=False` on optional parameters: 010a4ac173

(I looked into getting the defaults from click but couldn't find a nice non-hacky way)

Reply:
If you want to force people to pass parameters to `option()` then I think it should probably have that specified as a required parameter.  That will give the clearest hinting and error messages.  To mix this in with my other comments, this would also lead to a `.required` attribute on the metadata class and then explicit passing to `click.option()` via `required=metadata.required`.

Reply:
There's a difference between something being non-optional but being not required because it has a default and having something be optional without a default. However, I wanted to preserve the feeling of using `click.option` when using an option decorator so I guess having `required` be a mandatory arg in any scenario is against that ethos. Removed: e20cf6644f